container-engine-lib/lib/backend_impls/remote_context_backend/backend_creator.go

```diff
@@ -0,0 +1,95 @@
+package remote_context_backend
+
+import (
+	"fmt"
+	"github.com/docker/docker/client"
+	"github.com/kurtosis-tech/kurtosis/container-engine-lib/lib/backend_impls/docker/docker_kurtosis_backend/backend_creator"
+	"github.com/kurtosis-tech/kurtosis/container-engine-lib/lib/backend_interface"
+	"github.com/kurtosis-tech/stacktrace"
+	"github.com/sirupsen/logrus"
+	"os"
+	"path"
+)
+
+const (
+	urlProtocol = "tcp"
+
+	noTempDirPrefix    = ""
+	tempDirNamePattern = "kurtosis_backend_tls_*"
+	caFileName         = "ca.pem"
+	certFileName       = "cert.pem"
+	keyFileName        = "key.pem"
+	tlsFilesPerm       = 0644
+)
+
+func GetContextAwareKurtosisBackend(
+	remoteBackendConfig *KurtosisRemoteBackendConfig,
+	optionalApiContainerModeArgs *backend_creator.APIContainerModeArgs,
+) (backend_interface.KurtosisBackend, error) {
+	localDockerBackend, err := backend_creator.GetLocalDockerKurtosisBackend(optionalApiContainerModeArgs)
+	if err != nil {
+		return nil, stacktrace.Propagate(err, "Unable to build Kurtosis Backend connected to local Docker instance")
+	}
+
+	if remoteBackendConfig == nil {
+		logrus.Debugf("Instantiating a context aware backend with no remote backend config ends up returning" +
+			"a regular local Docker backend.")
+		return localDockerBackend, nil
+	}
+	remoteDockerClient, err := buildRemoteDockerClient(remoteBackendConfig)
+	if err != nil {
+		return nil, stacktrace.Propagate(err, "Error building client configuration for Docker remote backend")
+	}
+	kurtosisRemoteBackend, err := backend_creator.GetDockerKurtosisBackend(remoteDockerClient, optionalApiContainerModeArgs)
+	if err != nil {
+		return nil, stacktrace.Propagate(err, "Error building Kurtosis remote Docker backend")
+	}
+	return newRemoteContextKurtosisBackend(localDockerBackend, kurtosisRemoteBackend), nil
+}
+
+func buildRemoteDockerClient(remoteBackendConfig *KurtosisRemoteBackendConfig) (*client.Client, error) {
+	var clientOptions []client.Opt
+
+	// host and port option
+	url := fmt.Sprintf("%s://%s:%d", urlProtocol, remoteBackendConfig.Host, remoteBackendConfig.Port)
+	clientOptions = append(clientOptions, client.WithHost(url))
+
+	// TLS option if config is present
+	if tlsConfig := remoteBackendConfig.Tls; tlsConfig != nil {
+		tlsFilesDir, err := writeTlsConfigToTempDir(tlsConfig.Ca, tlsConfig.ClientCert, tlsConfig.ClientKey)
+		if err != nil {
+			return nil, stacktrace.Propagate(err, "Error building TLS configuration to connect to remote Docker backend")
+		}
+		tlsOpt := client.WithTLSClientConfig(
+			path.Join(tlsFilesDir, caFileName),
+			path.Join(tlsFilesDir, certFileName),
+			path.Join(tlsFilesDir, keyFileName))
+		clientOptions = append(clientOptions, tlsOpt)
+	}
+
+	// API version negotiation option
+	clientOptions = append(clientOptions, client.WithAPIVersionNegotiation())
```

Comment:
what does this do?

Reply:
It's a standard thing to instantiate a Docker Client. As I understand it you can force the client to talk to only a certain version of Docker, by forcing the version number. This is a more permissive flag where docker client and server enters a negotiation and if they consider they are both mutually compatible, they move forward, otherwise the client fails. Given that docker backward compatibilty is pretty good, this is pretty standard to use this.

FWIW this is what we currently use to build our docker clients